daemon/loop.go

```diff
@@ -23,6 +23,7 @@ import (
 const (
 	// Timeout for git operations we're prepared to abandon
 	gitOpTimeout = 15 * time.Second
+	syncSetName  = "git"
```

Comment:
I think that having a constant `syncSetName` can be dangerous since it's used as the label value to gather synced resources:

1. when running multiple `fluxd` instances on non-overlapping resources from different git repos, this will result in a fight to garbage-collect each other's resources.

2. It may also prove problematic once (if we ever) support multiple repos in a single fluxd instance. 

3. It will cause fluxd to remove all the resources in a git repo once we change it's configuration to another git repo with different resources. Intuitively this may be correct, but can be potentially disastrous and easy to miss by users.

As a solution we could append a hash of the git repo details to the label value. When doing this we need to make sure that we overwrite the label value to that of the repository currently used (mergo can be tricky). Alternatively, we can add the hash to the label key, but this can end up leaving garbage behind.

Reply:
Yes, it should be something representing the git {repo,branch}. That may leave stragglers if people change the config, but at least they will be easy to find.

Reply:
So, are you going to go for a different key per repo?

Reply:
hash({URL, branch}) (or even just {URL, branch}, used as the label value, I would think. But then, if you run two fluxd's and point them at different the same repo and branch, but different directories, they will garbage collect each other's resources.

The bounds are, I guess:
 - it's OK to have too much information, so long as it's stable between restarts (otherwise fluxd won't collect things it should collect).
 - having too little information in the label means you can (easily) have two fluxd's that conflict.
Augh..

Reply:
Uhm, I am thinking that, to be safe, we could stop guaranteeing resources being garbage collected between restarts. That is, if a resource is removed in git while fluxd is down, it won't be garbage collected.

That would allow us to use a label value (or key) associated with the current execution/instance of fluxd (e.g. through a random UUID generated at start). In this way, different fluxd instances would only be competing with each other if they are controlling exactly the same resource (which is unavoidable), but would never garbage collect each other's resources.

As an alternative, we could extend the hash to git directories and whatever we think is relevant to avoid mutual garbage collections.

Note that, even when using a git hash of the repo, we would miss garbage collections in restarts if the git repo is moved somewhere else.

Reply:
> we could stop guaranteeing resources being garbage collected between restarts

I think this would be a bit of a disaster for people operating fluxd -- since fluxd can be restarted arbitrarily, this means you'd never know whether garbage collection had been done, or wasn't (ever) going to be.

> As an alternative, we could extend the hash to git directories and whatever we think is relevant to avoid mutual garbage collections
> [...]
> I don't think we can really guarantee garbage collections between restarts in all cases without risking conflicts between fluxd instances.

Good insight: using the path (either git-path argument, or source path of the resource, or something along those lines) means it's more sensitive to changes in the configuration:

| Labelling | What happens with two fluxds | What happens after reconfiguration |
|-------|--------------------------------------|----------------------------------------|
| constant e.g., "git" | if fluxds are syncing different repos or directories, they will GC each other's resources | everything proceeds as before |
| git repo URL | if fluxds are given the same repo, but _different_ directories, they will GC each other's resources | if the git repo is moved, everything that's still there will get relabelled; anything that's now missing will be orphaned (never collected) |
| git URL and git-path or source path | fluxd's may relabel each other's resources if they are in common, but won't GC each other's resources | if you remove a directory from fluxd's config, it will be orphaned (never collected) |

We could use the `--git-label`, which already needs to be distinct per fluxd (though only if you're using the same repo). Using a unique ID:

| Labelling | What happens with two fluxds | What happens after reconfiguration |
|-------|--------------------------------------|----------------------------------------|
| fluxd ID | fluxds won't collect each other's resources; | if you change who is responsible for a resource, it'll get relabelled |

Reply:
> if you remove a directory from fluxd's config, it will be orphaned (never collected)

I don't think this is a terrible thing (particularly if you consider the possibility of making mistakes in configuration errors)

Reply:
After discussing it offline it was decided to go for a hash of the git URL + git paths